File: blasd/syr2k.c

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

#include "dtype.h"
#include "interfaces.h"


// SYR2K;
//   C = alpha*A*B.T + alpha*B*A.T + beta*C  
//   C = alpha*A.T*B + alpha*B.T*A + beta*C  if flags & GOMAS_TRANS
/*
 * Symmetric rank2 update
 *
 * upper
 *   C00 C01 C02    C00  C01 C02     A0               B0           
 *    0  C11 C12 =   0   C11 C12  +  A1 * B0 B1 B2 +  B1 * A0 A1 A2
 *    0   0  C22     0   0   C22     A2               B2           
 *
 * lower:
 *   C00  0   0    C00   0   0      A0               B0           
 *   C10 C11  0 =  C10  C11  0   +  A1 * B0 B1 B2 +  B1 * A0 A1 A2
 *   C20 C21 C22   C20  C21 C22     A2               B2           
 *
 */
void __rank2_blk(mdata_t *C, const mdata_t *A, const mdata_t *B,
                 DTYPE alpha, DTYPE beta, int flags,
                 int P, int S, int E,  int KB, int NB, int MB)
{
  register int i, j, nI, nC;
  mdata_t Cd, Ad, Bd;
  mdata_t Acpy, Bcpy;
  cache_t cache;
  DTYPE Abuf[MAX_KB*MAX_MB], Bbuf[MAX_KB*MAX_NB] __attribute__((aligned(64)));

  if (E-S <= 0 || P <= 0)
    return;

  if (NB > E-S) {
    NB = E-S;
  }

  Acpy = (mdata_t){Abuf, MAX_KB};
  Bcpy = (mdata_t){Bbuf, MAX_KB};
  cache = (cache_t){&Acpy, &Bcpy, KB, NB, MB};

  if (flags & GOMAS_TRANSA || flags & GOMAS_TRANS) {
    //   C = alpha*A.T*B + alpha*B.T*A + beta*C 
    for (i = S; i < E; i += NB) {
      nI = E - i < NB ? E - i : NB;
    
      // 1. update on diagonal
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, A, 0, i);
      __subblock(&Bd, B, 0, i);
      __rank_diag(&Cd, &Ad, &Bd, alpha, beta, flags, P, nI, &cache);

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Ad, A, 0, i);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Bd, B, 0, S);
        nC = E - i - nI;
      } else {
        __subblock(&Cd, C, i, i+nI);
        __subblock(&Bd, B, 0, i+nI);
        nC = i;
      }
      __gemm_colwise_inner_scale_c(&Cd, &Ad, &Bd, alpha, beta, GOMAS_TRANSA,
                                     P, 0, nC, 0, nI, &cache); 

      // 2nd part
      // 1. update on diagonal
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, A, 0, i);
      __subblock(&Bd, B, 0, i);
      __rank_diag(&Cd, &Bd, &Ad, alpha, 1.0, flags, P, nI, &cache);

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Bd, B, 0, i);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Ad, A, 0, S);
        nC = E - i - nI;
      } else {
        __subblock(&Cd, C, i, i+nI);
        __subblock(&Ad, A, 0, i+nI);
        nC = i;
      }
      __gemm_colwise_inner_scale_c(&Cd, &Bd, &Ad, alpha, 1.0, GOMAS_TRANSA,
                                     P, 0, nC, 0, nI, &cache);

    }
  } else {
    //   C = alpha*A*B.T + alpha*B*A.T + beta*C  
    for (i = S; i < E; i += NB) {
      nI = E - i < NB ? E - i : NB;
    
      // 1. update on diagonal
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, A, i, 0);
      __subblock(&Bd, B, i, 0);
      __rank_diag(&Cd, &Ad, &Bd, alpha, beta, flags, P, nI, &cache); 

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Ad, A, i, 0);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Bd, B, S, 0);
        nC = i;
      } else {
        __subblock(&Cd, C, i,    i+nI);
        __subblock(&Bd, B, i+nI, 0);
        nC = E - i - nI;
      }

      __gemm_colwise_inner_scale_c(&Cd, &Ad, &Bd, alpha, beta, GOMAS_TRANSB,
                                     P, 0, nC, 0, nI, &cache);

      // 1. update on diagonal
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, B, i, 0);
      __subblock(&Bd, A, i, 0);
      __rank_diag(&Cd, &Ad, &Bd, alpha, 1.0, flags, P, nI, &cache);

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Ad, B, i, 0);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Bd, A, S, 0);
        nC = i;
      } else {
        __subblock(&Cd, C, i,    i+nI);
        __subblock(&Bd, A, i+nI, 0);
        nC = E - i - nI;
      }

      __gemm_colwise_inner_scale_c(&Cd, &Ad, &Bd, alpha, 1.0, GOMAS_TRANSB,
                                   P, 0, nC, 0, nI, &cache);

    }
  }
}

// Local Variables:
// indent-tabs-mode: nil
// End:

File: blasd/syr2k.go

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

package blasd

// #cgo CFLAGS: -O3 -march=native -fomit-frame-pointer -ffast-math -Iinc -I.
// #cgo LDFLAGS: -lm
// #include "inc/interfaces.h"
import  "C"
import "unsafe"
import "github.com/hrautila/cmat"
import "github.com/hrautila/gomas"

func syr2k(Cc, Ac, Bc *cmat.FloatMatrix, alpha, beta float64, bits, P, S, E int, conf *gomas.Config)  {
    var Am, Cm, Bm C.mdata_t
    
    Am.md = (*C.double)(unsafe.Pointer(&Ac.Data()[0]))
    Am.step = C.int(Ac.Stride())
    Bm.md = (*C.double)(unsafe.Pointer(&Bc.Data()[0]))
    Bm.step = C.int(Bc.Stride())
    Cm.md = (*C.double)(unsafe.Pointer(&Cc.Data()[0]))
    Cm.step = C.int(Cc.Stride())

    C.__d_rank2_blk(
        (*C.mdata_t)(unsafe.Pointer(&Cm)), 
        (*C.mdata_t)(unsafe.Pointer(&Am)), 
        (*C.mdata_t)(unsafe.Pointer(&Bm)),
        C.double(alpha), C.double(beta), C.int(bits),
        C.int(P), C.int(S), C.int(E),
        C.int(conf.KB), C.int(conf.NB), C.int(conf.MB))
}


func Update2Sym(Cc, A, B *cmat.FloatMatrix, alpha, beta float64, bits int, confs... *gomas.Config)  *gomas.Error {

    conf := gomas.DefaultConf()
    if len(confs) > 0 {
        conf = confs[0]
    }

    ok := true
    cr, cc := Cc.Size()
    ar, ac := A.Size()
    br, bc := B.Size()

    if cr*cc == 0 {
        return nil
    }
    P := ac
    E := cr
    if bits & gomas.TRANS != 0 && bits & gomas.TRANSA == 0 {
        bits |= gomas.TRANSA
    }
    switch {
    case bits & gomas.TRANSA != 0:
        ok = cr == cc && cr == ac && bc == ac && br == ar
        P = ar
    default:
        ok = cr == cc && cr == ar && br == ar && bc == ac
    }
    if !ok {
        return gomas.NewError(gomas.ESIZE, "Update2Sym")
    }
    if conf.NProc == 1 || conf.WB <= 0 || E <= conf.WB {
        syr2k(Cc, A, B, alpha, beta, bits, P, 0, E, conf)
        return nil
    }
    // parallelized
    wait := make(chan int, 4)
    _, nN := blocking(0, E, conf.WB)
    nT := 0
    for j := 0; j < nN; j++ {
        jS := blockIndex(j, nN, conf.WB, E)
        jE := blockIndex(j+1, nN, conf.WB, E)
        task := func(q chan int) {
            syr2k(Cc, A, B, alpha, beta, bits, P, jS, jE, conf)
            q <- 1
        }
        conf.Sched.Schedule(gomas.NewTask(task, wait))
        nT += 1
    }
    for nT > 0 {
        <- wait
        nT -= 1
    }
    return nil
}

// Local Variables:
// tab-width: 4
// indent-tabs-mode: nil
// End:
